### PR TITLE
Add the post about PGA

### PR DESCRIPTION
**Pull request by vmarkovtsev**

### CODE REVIEW

**Comment by vmarkovtsev** [comment]
Deployed at staging as https://blog-staging.srcd.run/post/pga/

**Comment by eiso** [comment]
Instead of giving a review of the text, I actually think this whole post should be rewritten. It's telling a story of how it was created instead of an announcement. 

A good announcement IMO should be:

- What is it?
- For whom is this?
- Why is it valuable and why is it more valuable than the alternatives?
- How can you use it?

**Comment by vmarkovtsev** [comment]
Yep, this is not an announcement. I think Francesc wanted to make the announcement post, mine should be the second one or buried deep below the main content.

**Comment by campoy** [comment]
I agree with @eiso on this one. We should first write a blog post on what PGA is, rather than on how we got there.

I'm inclined to accompany it with a [dataset datasheet](https://arxiv.org/pdf/1803.09010.pdf).

**Comment by vmarkovtsev** [comment]
Neither this one was intended to published first. You are a bit weird guys. You definitely need to communicate more, at least answer (my) comments on GitHub and Slack and discuss your future plans with the rest of the team, instead of "shocking" ML and Applications teams once in a while. I discussed this issue with several people and we all agreed on that. I am desperately trying to initiate any discussion and the only way seems to be making a trolling PR like this one.

File: content/post/pga.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+date: 2018-04-10
+title: "Public Git Archive"
+image: /post/pga/logo.png
+description: "We have released a large dataset for MLonCode - Public Git Archive (PGA). It contains 182,000 top-starred repositories on GitHub and takes 3 TB of disk space. The paper which describes it was accepted
+to Mining Software Repositories conference. This post tells the PGA's story."
+categories: ["science", "technical"]
+---
+
+[![logo](/post/pga/logo.svg#left)
+Dataset landing](https://github.com/src-d/datasets/PublicGitArchive)
```
Comment (ajnavarro) [inline content/post/pga.md]: This link is wrong. It should be: https://github.com/src-d/datasets/tree/master/PublicGitArchive

Comment (vmarkovtsev) [reply content/post/pga.md]: Thanks for noticing!

**Comment by campoy** [comment]
Hey @vmarkovtsev,

It's never been my intention to shock the ML team at any point.
Please point me to your comments on GitHub or Slack to which I haven't answered and I'll gladly do so.

I wonder whether the shocking part of my work is to come up with crazy ideas, and what whether that bothers you.
This explorations are part of my responsibilities, coming up with crazy ideas that might not work, cool demos, etc. This is part of what Developer Relations does.

Regarding the tone of your comment, I detect some aggressiveness that I'd love to understand.
I'm looking forward to tomorrow's meeting on the topic.

**Comment by vmarkovtsev** [comment]
@campoy Let's decide about the destiny of this post. 20 days passed, we are still here without anything about PGA, whether it be an announcement or a how-was-made.

**Comment by campoy** [comment]
As agreed with @vmarkovtsev on Slack:

- May 1st: publish the go-license-detector blog post
- May 3rd: repost the go-license-detector blog post to Medium
- May 8th: publish a Public Git Archive blog post
- May 10th: republish PGA post to Medium
- May 15th: publish identifiers blog post + release dataset
- May 17th: republish to Medium

This post can be posted on May 22nd.
(We have too many blog posts now ... which is amazing problem to have!)

**Comment by vmarkovtsev** [comment]
@campoy @eiso Let's revise the schedule and make a new plan. We are now in a situation when PGA is presented in 3 days and our blog is dead silent.

**Comment by eiso** [comment]
@vmarkovtsev I discussed this with @campoy today and publishing on the blog is set for Thursday.

**Comment by campoy** [comment]
I think it's time to publish this blogpost, @vmarkovtsev!

Do you think it's ready for me to review? We could try to release it next Tuesday and ride the wave of popularity that the original PGA announcement got.

**Comment by campoy** [comment]
Is it ready for review, then?

**Comment by vmarkovtsev** [comment]
@campoy Done. I forgot to write which Monday haha.

**Comment by vmarkovtsev** [comment]
@vcoisne Tagging you just in case you feel like correcting stupid English mistakes.

**Comment by eiso** [comment]
I'll come back to this once it's gone through the first reviews by dev. rel. I feel this post can use some love & attention.

**Comment by campoy** [comment]
Hey @vmarkovtsev,

This is an interesting blog post, but I think it lacks structure  which makes it a tad confusing for the reader.
Would you mind coming up with a list of the points you want to make and how they relate to each other?

Also, who is the audience of this post? Details like what software was used to write the paper seem to indicate it is for other researchers, but the contents point more to people interested on using the dataset.

**Comment by vmarkovtsev** [comment]
I need some help here - completely out of resources for combing this. @vcoisne you offered help, didn't you?

File: content/post/pga_history.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+timestamps which allow identification. We've discovered that while the code itself may be OSS,
+its license does not extend to the development history. In other words,
+Git repositories distribution looks like a GDPR infringement. However, there is
+an [exclusion in GDPR](https://news.ycombinator.com/item?id=16509688) which really saves us:
+Git metadata is an inherent part of the dataset because anonymization changes the Git commit
+hashes. Besides, PGA does not contain private information in an easily accessible format because
+you still need to read the Git packfiles to extract it.
+
+GHTorrent is licensed under CC-BY-SA.
+
```
Comment (vcoisne) [inline content/post/pga_history.md]: I would remove the Paper, Poster, Cut scenes and Licensing sections above to shorten this blog post and focus on the retrieval problems & origin of PGA.

**Comment by vcoisne** [comment]
@vmarkovtsev do you still want to post this one ?

**Comment by campoy** [comment]
ping

**Comment by vcoisne** [comment]
@vmarkovtsev Do you think we can schedule this post for later this week ?

**Comment by vmarkovtsev** [comment]
https://github.com/src-d/blog/issues/256 pumped into my FIFO queue and preempted this.

**Comment by vmarkovtsev** [comment]
My bad, I meant LIFO of course :laughing:

**Comment by vmarkovtsev** [comment]
@campoy @vcoisne Wow that was faster and easier than I expected. Done. Please review again.

**Comment by campoy** [comment]
DCO is missing.

File: content/post/pga_history.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+for training Machine Learning on Source Code (aka MLonCode) models.
+It contains 182,000 top-starred repositories on GitHub and occupies 3 TB on disk.
+The repositories were cloned in February-March 2018.
+Check out the [announcement post](https://blog.sourced.tech/post/announcing-pga/) for more information.
+You should check out [Engine](https://github.com/src-d/engine) which allows
+to run SQL queries on top the PGA and do other cool things.
+
+## Origins
+
+We've wanted to release all the Git repositories we are able to clone since the source{d}'s beginning. Back in 2016 we stored millions of Git [packfiles](https://git-scm.com/book/en/v2/Git-Internals-Packfiles) in Google Cloud Storage and the metadata in MongoDB. It worked well for us but that scheme did not allow to share the data. Then we decided to go baremetal in 2017 and started to build our infrastructure from scratch. One of the
```
Comment (campoy) [inline content/post/pga_history.md]: s/baremetal/bare metal/

**Comment by vmarkovtsev** [comment]
@campoy not required for source{d} employees.

File: content/post/pga_history.md (hunk 3 above)
Comment (campoy) [reply content/post/pga_history.md]: s/the source{d}'s beginning/the beginning of source{d}/

File: content/post/pga_history.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
+---
+author: vadim
+date: 2018-10-30
+title: "The Public Git Archive Story"
+image: /post/pga_history/logo.png
+description: "We have recently released a large dataset for MLonCode - Public Git Archive (PGA). It contains 182,000 top-starred repositories on GitHub and occupies 3 TB on disk.  This post tells the story how PGA emerged: why, how, and what's next."
```
Comment (campoy) [inline content/post/pga_history.md]: s/occupies/takes/

File: content/post/pga_history.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+categories: ["science", "technical"]
+---
+
+## What is Public Git Archive?
+
+Public Git Archive is the result of months of efforts curating a dataset suitable
+for training Machine Learning on Source Code (aka MLonCode) models.
+It contains 182,000 top-starred repositories on GitHub and occupies 3 TB on disk.
```
Comment (campoy) [inline content/post/pga_history.md]: s/occupies/takes/

File: content/post/pga_history.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+The repositories were cloned in February-March 2018.
+Check out the [announcement post](https://blog.sourced.tech/post/announcing-pga/) for more information.
+You should check out [Engine](https://github.com/src-d/engine) which allows
+to run SQL queries on top the PGA and do other cool things.
+
+## Origins
+
+We've wanted to release all the Git repositories we are able to clone since the source{d}'s beginning. Back in 2016 we stored millions of Git [packfiles](https://git-scm.com/book/en/v2/Git-Internals-Packfiles) in Google Cloud Storage and the metadata in MongoDB. It worked well for us but that scheme did not allow to share the data. Then we decided to go baremetal in 2017 and started to build our infrastructure from scratch. One of the
+important changes was the creation of [Siva](https://github.com/src-d/go-siva) Git repository archive format.
```
Comment (campoy) [inline content/post/pga_history.md]: of *the* Siva

File: content/post/pga_history.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+Siva resembles [Tar](https://www.gnu.org/software/tar/) but the index is placed at the end of the file to allow
+cheap append operation. We re-invented the concept of a "rooted repository" where all branches from the same root are placed together, and therefore the corresponding packfiles are appended to the same Siva file while we are massively cloning Git repositories. It is not unique - GitHub has a similar approach to storing forks
```
Comment (campoy) [inline content/post/pga_history.md]: cheap append *operations*

File: content/post/pga_history.md [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+internally. The main benefit is re-using the Git objects between branches and thus reducing the size on disk.
```
Comment (campoy) [inline content/post/pga_history.md]: s/re-using/reusing/

File: content/post/pga_history.md [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+difficult to open access to the whole Git world (we clone from BitBucket, Savannah, servers in the wild, etc.)
+because the size is too big, so we started from putting a threshold to the number of stargazers on GitHub
+and picking the most popular projects. This threshold was set to **50** from Vadim's
+historical default limit in [GitHubStars](https://github.com/vmarkovtsev/githubstars).
+The plan was to distribute the Siva files generated from those popular repositories.
+Later we realized that having only the Siva files is not enough - people wanted to tighten the stars limit,
+filter by main programming language or by license. The idea of an index file appeared where we would place
+various metadata, either existing or mined after cloning.
+
+Cloning Git repositories is ~hard~ 🔥HELL🔥:
```
Comment (campoy) [inline content/post/pga_history.md]: hard *as* hell, maybe?

i'd avoid references to hell, but that's up to you 😄

File: content/post/pga_history.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+* Integration between Gerrit and GitHub produces hundreds of standalone Git references. E.g. [google/angle](https://github.com/google/angle) becomes spread over 6,600 Siva files, each with a tiny single ref coming from
+a Gerrit review.
+* [Git Bomb](https://github.com/Katee/git-bomb) - an "uncloneable" repository. By the way, we cloned it.
+* Some repositories have packfiles in the [old deprecated format]. go-git handles them.(https://github.com/git/git/blob/master/Documentation/technical/pack-format.txt#L39).
```
Comment (campoy) [inline content/post/pga_history.md]: missing link?

File: content/post/pga_history.md [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+* Load balancing is difficult. 90% of the repositories are compact and fly fast, however, the rest 10% occupy 50% of the dataset size and are damn slow to process.
```
Comment (campoy) [inline content/post/pga_history.md]: the remaining 10% take up 50% of the total size of the dataset and are really lengthy to process

File: content/post/pga_history.md [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+* Analyzing big repositories requires much memory, as much as 128 GB. Thus our pipeline kept exploding again and again and we restarted it with a firm hand.
+
+We had to take into account the removed repositories. Over 3,000 were deleted within few months.
+We discovered a few repositories with porno while inspecting the outliers by repository size (no links here, sorry, join our community Slack and direct message).
```
Comment (campoy) [inline content/post/pga_history.md]: s/porno/porn/

File: content/post/pga_history.md [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+<embed src="https://arxiv.org/pdf/1803.10144" width="600" height="600" alt="PGA paper from ArXiV">
+
+{{% caption src="/post/pga_history/presentation.jpg" %}}
+Vadim is presenting Public Git Archive on MSR'18/ICSE'18.
+{{% /caption %}}
+
+We used [Overleaf](https://www.overleaf.com/) to write the paper. When we reached the internal review
+phase, we pushed it to a private GitHub repository and the rest of the edits happenned in pull requests.
+Thus we avoided the "cold start" problem where the first pull request is huge and nightmare to read and collaborate. If we did not have to use TeX, we would use Google Docs instead.
```
Comment (campoy) [inline content/post/pga_history.md]: "is a huge nightmare", or "is huge and a nightmare"

File: content/post/pga_history.md [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+to the "Data Showcase" track on [MSR'18](http://2018.msrconf.org).
+We took action, wrote the paper and luckily for us it was accepted.
+
+<embed src="https://arxiv.org/pdf/1803.10144" width="600" height="600" alt="PGA paper from ArXiV">
+
+{{% caption src="/post/pga_history/presentation.jpg" %}}
+Vadim is presenting Public Git Archive on MSR'18/ICSE'18.
+{{% /caption %}}
+
+We used [Overleaf](https://www.overleaf.com/) to write the paper. When we reached the internal review
```
Comment (campoy) [inline content/post/pga_history.md]: This section is completely off topic, I'd drop it.

File: content/post/pga_history.md [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+{{% caption src="/post/pga_history/poster.jpg" %}}
+Conference attendees looking at the PGA poster on MSR'18/ICSE'18.
+{{% /caption %}}
+
+## Future
+
+We plan to release the next version of PGA later this year or early next year.
+It is included into our [Q4 OKRs](https://github.com/src-d/okrs/tree/2018Q4#release-a-new-version-of-pga-infraapps).
+Stay tuned.
```
Comment (campoy) [inline content/post/pga_history.md]: I would finish with a call to action. Maybe something regarding on how to learn more about PGA and start using it.

File: content/post/pga_history.md (hunk 4 above)
Comment (vmarkovtsev) [reply content/post/pga_history.md]: (applied the change)

But how do you explain https://www.google.es/search?q=%22occupy+disk+space%22

**Comment by vmarkovtsev** [comment]
@campoy All feedback addressed

File: content/post/pga_history.md (hunk 4 above)
Comment (campoy) [reply content/post/pga_history.md]: English is complicated 🤷‍♂️

**Comment by campoy** [comment]
Approved. @vcoisne to decide on publication date.

**Comment by vcoisne** [comment]
scheduled for 11/14 @vmarkovtsev

**Pull request merged**